subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java

```diff
@@ -47,20 +47,20 @@
 import java.io.File;
 import java.util.Optional;
 
-public class CaptureStateBeforeExecutionStep extends BuildOperationStep<PreviousExecutionContext, CachingResult> {
+public class CaptureStateBeforeExecutionStep<C extends PreviousExecutionContext, R extends CachingResult> extends BuildOperationStep<C, R> {
```

Comment:
Consistency between steps, and the ability to better move the steps around if need be.

Reply:
I don't think consistency is a good reason. We have been removing and adding generics for some time. Maybe the generics are wrong and we are missing something here?

It seems to me that there are different kind of steps:
- Steps which run stuff before -> Context of the delegate can be more, or can be the same, Result doesn't matter
- Steps which run stuff after -> Why does the context even matter here, they all should have the full context, right? They may change the result though.
- Steps which do actual work, aka produce a result.

Maybe we should go into this direction instead of trying to fit everything into the `Step` concept with somewhat arbitrary generics?

I don't feel very strongly about doing this now.

Reply:
For now this helps remove unnecessary complexity when moving the steps around. A better solution will be when we split the steps into context building / result processing, see below.